rootchain/sampool/pool.go

```diff
@@ -0,0 +1,146 @@
+package sampool
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/0xPolygon/polygon-edge/rootchain"
+)
+
+var (
+	ErrInvalidHash      = errors.New("invalid SAM hash")
+	ErrInvalidSignature = errors.New("invalid SAM signature")
+	ErrStaleMessage     = errors.New("stale SAM received")
+)
+
+type Verifier interface {
+	VerifyHash(rootchain.SAM) error
+	VerifySignature(rootchain.SAM) error
+
+	Quorum(uint64) bool
+}
+
+type SAMPool struct {
+	verifier Verifier
+
+	mux      sync.Mutex
+	messages map[uint64]samBucket
+
+	lastProcessedIndex uint64
+}
+
+func New(verifier Verifier) *SAMPool {
+	return &SAMPool{
+		verifier: verifier,
+		mux:      sync.Mutex{},
+		messages: make(map[uint64]samBucket),
+	}
+}
+
+func (p *SAMPool) AddMessage(msg rootchain.SAM) error {
+	if err := p.verifySAM(msg); err != nil {
+		return err
+	}
+
+	p.addSAM(msg)
+
+	return nil
+}
+
+func (p *SAMPool) Prune(index uint64) {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	for idx := range p.messages {
+		if idx <= index {
+			delete(p.messages, idx)
+		}
+	}
+
+	p.lastProcessedIndex = index
+}
+
+// TODO: Peek or Pop might be redundant
+func (p *SAMPool) Peek() rootchain.VerifiedSAM {
+	p.mux.Lock()
+	defer p.mux.Unlock()
+
+	expectedIndex := p.lastProcessedIndex + 1
+
+	bucket := p.messages[expectedIndex]
+	if bucket == nil {
+		return nil
+	}
+
+	messages := bucket.getQuorumMessages(p.verifier.Quorum)
+	if len(messages) == 0 {
+		return nil
+	}
+
+	result := make([]rootchain.SAM, len(messages))
+	copy(result, messages)
```

Comment:
Can you explain why shallow copy is necessary here?

Reply:
To protect against race conditions.

In an unlikely event that a call to `Peek` is made while there is a call to `AddMessage`, the original [slice reference](https://github.com/0xPolygon/polygon-edge/blob/7e2a9c397288c1edcb2c769c2a5d38ecd4430330/rootchain/sampool/bucket.go#L12) would be exposed for both read and write. The `copy` call is there just to ensure no 2 goroutines will be mutating this slice at the same time.

Reply:
Why mux can't prepend this?

Reply:
If I return the underlying slice, then the caller can do a read on it while an event being added in the backround is aboir to write to it